h5py/_hl/dataset.py

```diff
@@ -336,61 +338,72 @@ def __init__(self, dset, source_sel=None):
+                for dim in range(self._rank)
             )
         else:
             if isinstance(source_sel, slice):
                 self._sel = (source_sel,)
             else:
                 self._sel = source_sel
-        if len(self._sel) != rank:
+        if len(self._sel) != self._rank:
             raise ValueError("Invalid selection - selection region must have same rank as dataset")
         self._chunk_index = []
-        for dim in range(rank):
+        for dim in range(self._rank):
             s = self._sel[dim]
             if s.start < 0 or s.stop > self._shape[dim] or s.stop <= s.start:
                 raise ValueError("Invalid selection - selection region must be within dataset space")
             index = s.start // self._layout[dim]
             self._chunk_index.append(index)
 
+
     def __iter__(self):
         return self
 
     def __next__(self):
-        rank = len(self._shape)
-        slices = []
-        if rank == 0 or self._chunk_index[0] * self._layout[0] >= self._sel[0].stop:
+        if self._rank == 0 or self._chunk_index[0] * self._layout[0] >= self._sel[0].stop:
             # ran past the last chunk, end iteration
             raise StopIteration()
 
-        for dim in range(rank):
-            s = self._sel[dim]
-            start = self._chunk_index[dim] * self._layout[dim]
-            stop = (self._chunk_index[dim] + 1) * self._layout[dim]
-            # adjust the start if this is an edge chunk
-            if start < s.start:
-                start = s.start
-            if stop > s.stop:
-                stop = s.stop  # trim to end of the selection
-            s = slice(start, stop, 1)
-            slices.append(s)
+        for rank in range(self._rank):
+            self._slices[rank] = self._calculate_next_chunk(rank)
 
         # bump up the last index and carry forward if we run outside the selection
-        dim = rank - 1
+        dim = self._rank - 1
         while dim >= 0:
             s = self._sel[dim]
             self._chunk_index[dim] += 1
 
             chunk_end = self._chunk_index[dim] * self._layout[dim]
             if chunk_end < s.stop:
+                chunk_sel = self._slices[dim]
+                # there's a chance that our calculated chunk slice isn't a subset of
+                # the requested  slice. Check the bounds and increment further if so
+                for edge in (chunk_sel.start, chunk_sel.stop):
+                    if not (s.start <= edge <= s.stop):
+                        self._chunk_index[dim] = 1
+                        self._slices[dim] = self._calculate_next_chunk(dim)
+                        self._chunk_index[dim] += 1
+                        break
                 # we still have room to extend along this dimensions
-                return tuple(slices)
+                break
 
             if dim > 0:
                 # reset to the start and continue iterating with higher dimension
                 self._chunk_index[dim] = 0
```

Comment:
Looking at the code in master at present, this is the bit I'd be looking to change. When it hits the end of the selection in one dimension, this is telling it to go back to the start in that dimension - but the overall start of the dimension rather than the start of the selection. So I would try something like this:

```suggestion
                self._chunk_index[dim] = s.start // self._layout[dim]
```

If that works, I suspect it might fix it without the added complexity elsewhere in the class. In particular, I'm suspicious of the line `self._chunk_index[dim] = 1` you've added a few lines above. I suspect this hardcoded 1 works for the particular case you've added as a test, but will still be wrong in other cases.

Reply:
I will get this tomorrow. I hope it's simpler bc I felt what I was doing was complicated but it was only solution I was conceptualizing.

Reply:
`git checkout master -- h5py/_hl/dataset.py` and swapping that does indeed make it pass for the additional test case.

This is good because I was uncomfortable with the slowness I introduced in the common case.